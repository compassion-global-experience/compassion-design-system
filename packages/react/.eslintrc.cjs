module.exports = {
  extends: [
    'standard-jsx',
    'standard-react',
    'plugin:jsx-a11y/recommended',
    'plugin:react/recommended',
    'plugin:react/jsx-runtime',
    'plugin:react-hooks/recommended',
    'plugin:storybook/recommended',
  ],
  env: {
    browser: true,
    es6: true,
    jest: true,
    node: false,
  },
  rules: {
    'jsx-quotes': ['error', 'prefer-double'],
  },
};
