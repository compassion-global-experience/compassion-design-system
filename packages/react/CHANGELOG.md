# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.1.0 (2023-04-08)

### Features

- **react:** css usage and documentation ([11e317c](https://github.com/compassion-global-experience/compassion-design-system/commit/11e317cca4ecdaa385e86d3a5571d7dca554685d))
- **react:** re-export css package ([67549a5](https://github.com/compassion-global-experience/compassion-design-system/commit/67549a58de3de6e02b58c5bdb6f4605001aae783))
- **react:** switch between themes in storybook ([79119ba](https://github.com/compassion-global-experience/compassion-design-system/commit/79119ba48ed3410eaebe9a91ccd076b05c2580c1))
