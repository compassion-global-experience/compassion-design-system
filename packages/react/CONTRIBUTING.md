# Contributing to the `react-web` Package

The `react` package is structured with a component-per-directory convention
inside the `src` directory. Each component directory generally contains 3 items:

- A **TSX file** that describes the component‚Äôs structure and behavior
- A **.stories.ts file** that illustrates most of the ways the component can
  be configured
- An **index.ts** file that exports the TSX component‚Äôs contents

## Workflow

The package uses [Storybook](https://storybook.js.org/docs/react/get-started/introduction)
to facilitate component development and documentation. Storybook uses the files
mentioned above to render components in their various states, and serves as
documentation of those components by surfacing code comments, `propTypes`, and
`defaultProps` values.

To launch Storybook, make sure you‚Äôre in the [/packages/react](/packages/react)
directory and run the command `npm run storybook`. After the project is built,
Storybook will launch in your default browser. New components or modifications
will be displayed in real time! üéâ

The configurations outlined in each component‚Äôs .stories.ts file can be seen in
Storybook‚Äôs **Docs** tab.

## Commit Messages & Versioning

Every package in the compassion-gds monorepo adheres to
[semantic versioning](https://semver.org/). In order to simplify the versioning
and release process, please be sure your commit messages adhere to the
[‚ÄúConventional Commits‚Äù pattern](https://www.conventionalcommits.org/en/v1.0.0/#summary).
You may wish to install the [commitizen/cz-cli](https://github.com/commitizen/cz-cli)
to make this near-bulletproof.

## Making a Pull Request

Once your changes are ready, open a pull request against the `development`
branch of the project, and it will be merged in as quickly as possible!
