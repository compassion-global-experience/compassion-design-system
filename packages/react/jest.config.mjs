/*
 * For a detailed explanation regarding each configuration property, visit:
 * https://jestjs.io/docs/configuration
 */

export default {
  // A preset that is used as a base for Jest's configuration
  preset: 'ts-jest/presets/default-esm',

  // Automatically clear mock calls, instances, contexts and results before every test
  clearMocks: true,

  // Indicates whether the coverage information should be collected while executing the test
  collectCoverage: true,

  // The directory where Jest should output its coverage files
  coverageDirectory: "coverage",

  // Indicates which provider should be used to instrument code for coverage
  coverageProvider: "v8",

  // The test environment that will be used for testing
  testEnvironment: "jsdom",

  // A map from regular expressions to module names or to arrays of module names that allow to stub out resources with a single module
  moduleNameMapper: {
    ".(css|less|scss)$": "identity-obj-proxy",
  },

  // Whether to use watchman for file crawling
  watchman: true,
};
