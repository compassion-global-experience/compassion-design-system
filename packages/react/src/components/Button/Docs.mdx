import { Markdown } from '@storybook/blocks';
import {
  Canvas,
  Controls,
  Meta,
  Stories,
  Story,
  Title,
} from '@storybook/addon-docs';

import * as ButtonStories from './Button.stories.tsx';

import _01_Overview from '@compassion-design-system/core/src/components/Button/docs/01_Overview.md';
import _02_Usage from '@compassion-design-system/core/src/components/Button/docs/02_Usage.md';
import _03_Kind from '@compassion-design-system/core/src/components/Button/docs/03_Kind.md';
import _04_Emphasis from '@compassion-design-system/core/src/components/Button/docs/04_Emphasis.md';
import _05_Size from '@compassion-design-system/core/src/components/Button/docs/05_Size.md';
import _06_State from '@compassion-design-system/core/src/components/Button/docs/06_State.md';
import _07_OtherProperties from '@compassion-design-system/core/src/components/Button/docs/07_OtherProperties.md';
import _08_BestPractices from '@compassion-design-system/core/src/components/Button/docs/08_BestPractices.md';

<Meta of={ButtonStories} />

<Title />

<Markdown>{_01_Overview}</Markdown>

<Canvas of={ButtonStories.Default} sourceState="shown" withToolbar="true" />
<Controls of={ButtonStories.Default} />

<Markdown>{_02_Usage}</Markdown>

## Component Properties

<Markdown>{_03_Kind}</Markdown>
<Canvas of={ButtonStories.Kinds} />

<Markdown>{_04_Emphasis}</Markdown>
<Canvas of={ButtonStories.Emphasis} />

<Markdown>{_05_Size}</Markdown>
<Canvas of={ButtonStories.Size} />

<Markdown>{_06_State}</Markdown>
<Canvas of={ButtonStories.State} />

<Markdown>{_07_OtherProperties}</Markdown>
<Canvas of={ButtonStories.WithIcon} />

<Markdown>{_08_BestPractices}</Markdown>
