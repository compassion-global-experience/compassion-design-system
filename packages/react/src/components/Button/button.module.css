@import "~@compassion-gds/css/src/components/Button/button.module.css";
