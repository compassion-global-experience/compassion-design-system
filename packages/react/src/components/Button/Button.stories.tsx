import { Meta } from '@storybook/react';
import Button from './Button';
import { ArrowLeft, ArrowRight, Info } from 'phosphor-react';

const meta: Meta<typeof Button> = {
  /* ðŸ‘‡ The title prop is optional.
   * See https://storybook.js.org/docs/7.0/react/configure/overview#configure-story-loading
   * to learn how to generate automatic titles
   */
  title: 'Components/Button',
  component: Button,
};

export default meta;

// More on args: https://storybook.js.org/docs/react/writing-stories/args
export const Primary = {
  args: {
    mode: 'primary',
    label: 'Button',
  },
};

export const Secondary = {
  args: {
    mode: 'secondary',
    label: 'Button',
  },
};

export const Tertiary = {
  args: {
    mode: 'tertiary',
    label: 'Button',
  },
};

export const Small = {
  args: {
    size: 'small',
    label: 'Button',
  },
};

export const Medium = {
  args: {
    size: 'medium',
    label: 'Button',
  },
};

export const Large = {
  args: {
    size: 'large',
    label: 'Button',
  },
};

export const Disabled = {
  args: {
    disabled: true,
    label: 'Button',
  },
};

export const WithIconRight: typeof meta = {
  args: {
    mode: 'primary',
  },
  render: (args) => (
    <Button {...args}>
      Log in <ArrowRight />
    </Button>
  ),
};

export const WithIconLeft: typeof meta = {
  args: {
    mode: 'primary',
  },
  render: (args) => (
    <Button {...args}>
      <ArrowLeft /> Go back
    </Button>
  ),
};

export const WithIconOnBothSides: typeof meta = {
  args: {
    mode: 'primary',
  },
  render: (args) => (
    <Button {...args}>
      <ArrowLeft /> Options <ArrowRight />
    </Button>
  ),
};

export const IconOnly = {
  args: {
    mode: 'primary',
    size: 'small',
    label: <Info size={24} />,
  },
};
