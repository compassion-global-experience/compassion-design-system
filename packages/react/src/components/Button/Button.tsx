'use client';

import { MouseEvent, ButtonHTMLAttributes, forwardRef } from 'react';
import styles from '@compassion-design-system/core/src/components/Button/button.module.css';
import { getClasses } from '../../utils/classes';
import { cdsIconMap } from '../Icon/Icon';

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  label: string;
  showLabel?: boolean;
  kind?: 'cta' | 'white' | 'inverted' | 'destructive';
  emphasis?: 'primary' | 'secondary' | 'tertiary';
  size?: 'small' | 'medium' | 'large';
  disabled?: boolean;
  icon?: keyof typeof cdsIconMap;
  iconDescription?: string;
  iconPosition?: 'start' | 'end';
  onClick?: (event: MouseEvent<HTMLButtonElement>) => void;
}

const Button = forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => {
  const {
    label,
    showLabel = true,
    kind,
    emphasis = 'primary',
    size = 'medium',
    disabled = false,
    className,
    style,
    onClick,
    icon,
    iconDescription,
    iconPosition,
    ...rest
  } = props;

  const IconComponent = cdsIconMap[icon];

  const buttonSizeToIconSizeMap: { [key: string]: number } = {
    small: 13,
    medium: 16,
    large: 20,
  };

  const ButtonIcon = () => {
    return (
      <div className={getClasses(styles, buttonIconClassList)}>
        <IconComponent
          className=""
          alt={iconDescription}
          size={buttonSizeToIconSizeMap[size]}
        />
      </div>
    );
  };

  const buttonClassList = [`cds-button--${emphasis}`, `cds-button--${size}`];

  if (kind) {
    buttonClassList.push(`cds-button--${kind}`);
  }

  const buttonIconClassList = ['cds-button__icon'];

  if (IconComponent && showLabel) {
    if (iconPosition && iconPosition === 'end') {
      buttonClassList.push('cds-button--reverseTextIcon');
      buttonIconClassList.push('cds-button__icon--end');
    } else {
      buttonIconClassList.push('cds-button__icon--start');
    }
  }

  const buttonClassNames = getClasses(styles, buttonClassList, className);

  return (
    <button
      onClick={onClick}
      className={buttonClassNames}
      disabled={disabled}
      style={style}
      ref={ref}
      {...rest}
    >
      {IconComponent && <ButtonIcon />}
      {showLabel && label}
    </button>
  );
});

Button.displayName = 'Button';

export default Button;
