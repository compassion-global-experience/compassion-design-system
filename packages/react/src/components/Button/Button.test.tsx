import { render } from '@testing-library/react';

import Button from './Button';

describe('Button', () => {
  it('should render `Primary` mode with `Medium` size', () => {
    const { container } = render(<Button label="Click Me!" mode="primary" />);
    expect(container).toMatchSnapshot();
  });

  it('should render `Secondary` mode with `Small` size', () => {
    const { container } = render(<Button label="Click Me!" mode="secondary" size="small" />);
    expect(container).toMatchSnapshot();
  });

  it('should render `Tertiary` mode with `Large` size', () => {
    const { container } = render(<Button label="Click Me!" mode="tertiary" size="large" />);
    expect(container).toMatchSnapshot();
  });

  it('should render `Disabled` variant', () => {
    const { container } = render(<Button label="Disabled" mode="primary" disabled />);
    expect(container).toMatchSnapshot();
  });
});
