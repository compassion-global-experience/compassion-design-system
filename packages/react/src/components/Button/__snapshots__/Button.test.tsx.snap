// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render \`Disabled\` variant 1`] = `
<div>
  <button
    class="button-primary medium"
    disabled=""
  >
    Disabled
  </button>
</div>
`;

exports[`Button should render \`Primary\` mode with \`Medium\` size 1`] = `
<div>
  <button
    class="button-primary medium"
  >
    Click Me!
  </button>
</div>
`;

exports[`Button should render \`Secondary\` mode with \`Small\` size 1`] = `
<div>
  <button
    class="button-secondary small"
  >
    Click Me!
  </button>
</div>
`;

exports[`Button should render \`Tertiary\` mode with \`Large\` size 1`] = `
<div>
  <button
    class="button-tertiary large"
  >
    Click Me!
  </button>
</div>
`;
